File: CustomApplications/CSharp/GraphicsHowTo/DistanceOverlay.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Windows.Forms;
using System.Drawing;
using AGI.STKGraphics;

namespace GraphicsHowTo
{
    public class DistanceOverlay : StatusOverlay<double>
    {
        public DistanceOverlay(IAgStkGraphicsScene scene, IAgStkGraphicsSceneManager manager)
            :
            base(true, 0, 10000000, "0", "10000", manager)
        {
            m_Scene = scene;
            m_SceneManager = manager;
            ((AgStkGraphicsScene)scene).Rendering += new IAgStkGraphicsSceneEvents_RenderingEventHandler(Scene_Rendering);
        }

        void Scene_Rendering(object Sender, IAgStkGraphicsRenderingEventArgs Args)
        {
            Update(Value, m_SceneManager);
        }

        public override double ValueTransform(double value)
        {
            return (value >= 1) ? Math.Log10(value / 10000.0) : value;
        }

        public override double Value
        {
            get { return VectorMagnitude((double)m_Scene.Camera.Position.GetValue(0), (double)m_Scene.Camera.Position.GetValue(1), (double)m_Scene.Camera.Position.GetValue(2)); }
        }

        public override string Text
        {
            get { return "Current Distance:\n" + 
                String.Format(System.Globalization.CultureInfo.InvariantCulture, "{0,10:0.000} km", Value / 1000); }
        }

        /// <summary>
        /// Add a list of intervals of doubles using a different color for each.
        /// </summary>
        /// <param name="intervals">Collection of raw (non-ValueTransformed) Intervals.</param>
        internal void AddIntervals(ICollection<Interval> intervals)
        {
            Color[] colors = new Color[] 
            { 
                System.Drawing.Color.SkyBlue, System.Drawing.Color.LightGreen,
                System.Drawing.Color.Yellow, System.Drawing.Color.LightSalmon,
                System.Drawing.Color.DarkRed, System.Drawing.Color.MediumPurple 
            };

            int i = 0;
            foreach (Interval interval in intervals)
            {
                Indicator.AddInterval(ValueTransform(interval.Minimum), ValueTransform(interval.Maximum), 
                    IndicatorStyle.Bar, colors[i], m_SceneManager);
                i = (i + 1) % colors.Length;
            }
        }

        internal void RemoveIntervals(ICollection<Interval> intervals)
        {
            foreach (Interval interval in intervals)
            {
                Indicator.RemoveInterval(ValueTransform(interval.Minimum), ValueTransform(interval.Maximum));
            }
        }

        private static double VectorMagnitude(double x, double y, double z)
        {
            return Math.Sqrt((x * x) + (y * y) + (z * z));
        }

        private IAgStkGraphicsScene m_Scene;
        private IAgStkGraphicsSceneManager m_SceneManager;
    }
}

File: CustomApplications/VB.Net/GraphicsHowTo/DistanceOverlay.vb
Imports System.Collections.Generic
Imports System.Text
Imports System.Windows.Forms
Imports System.Drawing
Imports AGI.STKGraphics

Public Class DistanceOverlay
	Inherits StatusOverlay(Of Double)
	Public Sub New(scene As IAgStkGraphicsScene, manager As IAgStkGraphicsSceneManager)
		MyBase.New(True, 0, 10000000, "0", "10000", manager)
		m_Scene = scene
		m_SceneManager = manager
        AddHandler DirectCast(scene, AgStkGraphicsScene).Rendering, AddressOf Scene_Rendering
	End Sub

	Private Sub Scene_Rendering(Sender As Object, Args As IAgStkGraphicsRenderingEventArgs)
		Update(Value, m_SceneManager)
	End Sub

	Public Overrides Function ValueTransform(value As Double) As Double
		Return If((value >= 1), Math.Log10(value / 10000.0), value)
	End Function

	Public Overrides ReadOnly Property Value() As Double
		Get
			Return VectorMagnitude(CDbl(m_Scene.Camera.Position.GetValue(0)), CDbl(m_Scene.Camera.Position.GetValue(1)), CDbl(m_Scene.Camera.Position.GetValue(2)))
		End Get
	End Property

	Public Overrides ReadOnly Property Text() As String
		Get
            Return "Current Distance:" & vbLf & _
                    String.Format(System.Globalization.CultureInfo.InvariantCulture, "{0,10:0.000} km", Value / 1000)
		End Get
	End Property

	''' <summary>
	''' Add a list of intervals of doubles using a different color for each.
	''' </summary>
	''' <param name="intervals">Collection of raw (non-ValueTransformed) Intervals.</param>
	Friend Sub AddIntervals(intervals As ICollection(Of Interval))
        Dim colors As Color() = New Color() _
        { _
            System.Drawing.Color.SkyBlue, System.Drawing.Color.LightGreen, _
            System.Drawing.Color.Yellow, System.Drawing.Color.LightSalmon, _
            System.Drawing.Color.DarkRed, System.Drawing.Color.MediumPurple _
        }

		Dim i As Integer = 0
		For Each interval As Interval In intervals
            Indicator.AddInterval(ValueTransform(interval.Minimum), ValueTransform(interval.Maximum), IndicatorStyle.Bar, _
                                  colors(i), m_SceneManager)
			i = (i + 1) Mod colors.Length
		Next
	End Sub

	Friend Sub RemoveIntervals(intervals As ICollection(Of Interval))
		For Each interval As Interval In intervals
			Indicator.RemoveInterval(ValueTransform(interval.Minimum), ValueTransform(interval.Maximum))
		Next
	End Sub

	Private Shared Function VectorMagnitude(x As Double, y As Double, z As Double) As Double
		Return Math.Sqrt((x * x) + (y * y) + (z * z))
	End Function

	Private m_Scene As IAgStkGraphicsScene
	Private m_SceneManager As IAgStkGraphicsSceneManager
End Class
